File: rust/src/json_io.rs
use std::fs;
use std::io::Write;

use crate::simulate::SimulationResult;

// ============================================================================
// Parameter parsing
// ============================================================================

#[derive(Debug, Clone)]
pub struct SimParams {
    pub n: usize,
    pub vmode: f64,
    pub alpha_l: f64,
    pub alpha_r: f64,
    pub lambda: f64,
    pub t: f64,
    pub eta: f64,
    pub vg: f64,
    pub tau: f64,
    pub vsd_start: f64,
    pub vsd_end: f64,
    pub vsd_step: f64,
}

pub fn parse_params_json(filepath: &str) -> Option<SimParams> {
    let text = fs::read_to_string(filepath).ok()?;
    let root: serde_json::Value = serde_json::from_str(&text).ok()?;

    let params = root.get("parameters")?;
    let sweep = root.get("bias_sweep")?;

    let tau_val = match params.get("tau")? {
        serde_json::Value::String(s) if s == "Inf" => f64::INFINITY,
        serde_json::Value::Number(n) => n.as_f64().unwrap_or(f64::INFINITY),
        _ => f64::INFINITY,
    };

    Some(SimParams {
        n: params.get("N")?.as_u64()? as usize,
        vmode: params.get("vmode")?.as_f64()?,
        alpha_l: params.get("alphaL")?.as_f64()?,
        alpha_r: params.get("alphaR")?.as_f64()?,
        lambda: params.get("lambda")?.as_f64()?,
        t: params.get("T")?.as_f64()?,
        eta: params.get("eta")?.as_f64()?,
        vg: params.get("Vg")?.as_f64()?,
        tau: tau_val,
        vsd_start: sweep.get("Vsd_start")?.as_f64()?,
        vsd_end: sweep.get("Vsd_end")?.as_f64()?,
        vsd_step: sweep.get("Vsd_step")?.as_f64()?,
    })
}

// ============================================================================
// MATLAB reference loading
// ============================================================================

pub fn load_matlab_vsd(filepath: &str) -> Option<Vec<f64>> {
    let text = fs::read_to_string(filepath).ok()?;
    let root: serde_json::Value = serde_json::from_str(&text).ok()?;
    let arr = root.get("Vsd")?.as_array()?;
    Some(arr.iter().filter_map(|v| v.as_f64()).collect())
}

pub struct MatlabReference {
    pub i_tol: Vec<f64>,
    pub i_seq: Vec<f64>,
    pub i_cot: Vec<f64>,
}

pub fn load_matlab_reference(filepath: &str) -> Option<MatlabReference> {
    let text = fs::read_to_string(filepath).ok()?;
    let root: serde_json::Value = serde_json::from_str(&text).ok()?;

    let i_tol: Vec<f64> = root
        .get("I_tol")?
        .as_array()?
        .iter()
        .filter_map(|v| v.as_f64())
        .collect();
    let i_seq: Vec<f64> = root
        .get("I_seq")?
        .as_array()?
        .iter()
        .filter_map(|v| v.as_f64())
        .collect();
    let i_cot: Vec<f64> = root
        .get("I_cot")?
        .as_array()?
        .iter()
        .filter_map(|v| v.as_f64())
        .collect();

    Some(MatlabReference { i_tol, i_seq, i_cot })
}

// ============================================================================
// Output writing
// ============================================================================

fn write_json_array(f: &mut fs::File, name: &str, arr: &[f64], comma: bool) -> std::io::Result<()> {
    write!(f, "  \"{}\": [\n", name)?;
    for (i, &v) in arr.iter().enumerate() {
        if i + 1 < arr.len() {
            writeln!(f, "    {},", v)?;
        } else {
            writeln!(f, "    {}", v)?;
        }
    }
    if comma {
        writeln!(f, "  ],")?;
    } else {
        writeln!(f, "  ]")?;
    }
    Ok(())
}

pub fn write_results_json(
    filepath: &str,
    spec: &str,
    params: &SimParams,
    result: &SimulationResult,
    wall_time: f64,
) -> std::io::Result<()> {
    let mut f = fs::File::create(filepath)?;

    // Get timestamp
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();
    // Simple timestamp formatting (UTC)
    let secs_per_day = 86400u64;
    let secs_per_hour = 3600u64;
    let secs_per_min = 60u64;

    // Days since epoch to Y-M-D
    let mut days = now / secs_per_day;
    let time_of_day = now % secs_per_day;
    let hour = time_of_day / secs_per_hour;
    let minute = (time_of_day % secs_per_hour) / secs_per_min;
    let second = time_of_day % secs_per_min;

    // Simple year/month/day calculation
    let mut year = 1970i32;
    loop {
        let days_in_year = if year % 4 == 0 && (year % 100 != 0 || year % 400 == 0) {
            366
        } else {
            365
        };
        if days < days_in_year {
            break;
        }
        days -= days_in_year;
        year += 1;
    }
    let leap = year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
    let days_in_months: [u64; 12] = [
        31,
        if leap { 29 } else { 28 },
        31, 30, 31, 30, 31, 31, 30, 31, 30, 31,
    ];
    let mut month = 1u32;
    for &dim in &days_in_months {
        if days < dim {
            break;
        }
        days -= dim;
        month += 1;
    }
    let day = days + 1;

    let timestamp = format!(
        "{:04}-{:02}-{:02} {:02}:{:02}:{:02}",
        year, month, day, hour, minute, second
    );

    let rust_version = env!("CARGO_PKG_VERSION");

    writeln!(f, "{{")?;
    writeln!(f, "  \"spec\": \"{}\",", spec)?;
    writeln!(f, "  \"language\": \"Rust\",")?;
    writeln!(f, "  \"version\": \"rustc (franck-condon {})\",", rust_version)?;
    writeln!(f, "  \"wall_time_seconds\": {},", wall_time)?;
    writeln!(f, "  \"parameters\": {{")?;
    writeln!(f, "    \"N\": {},", params.n)?;
    writeln!(f, "    \"vmode\": {},", params.vmode)?;
    writeln!(f, "    \"alphaL\": {},", params.alpha_l)?;
    writeln!(f, "    \"alphaR\": {},", params.alpha_r)?;
    writeln!(f, "    \"lambda\": {},", params.lambda)?;
    writeln!(f, "    \"T\": {},", params.t)?;
    writeln!(f, "    \"eta\": {},", params.eta)?;
    writeln!(f, "    \"Vg\": {},", params.vg)?;
    if params.tau.is_infinite() {
        writeln!(f, "    \"tau\": \"Inf\"")?;
    } else {
        writeln!(f, "    \"tau\": {}", params.tau)?;
    }
    writeln!(f, "  }},")?;
    writeln!(f, "  \"bias_sweep\": {{")?;
    writeln!(f, "    \"Vsd_start\": {},", params.vsd_start)?;
    writeln!(f, "    \"Vsd_end\": {},", params.vsd_end)?;
    writeln!(f, "    \"Vsd_step\": {}", params.vsd_step)?;
    writeln!(f, "  }},")?;

    write_json_array(&mut f, "Vsd", &result.vsd, true)?;
    write_json_array(&mut f, "I_tol", &result.i_tol, true)?;
    write_json_array(&mut f, "I_seq", &result.i_seq, true)?;
    write_json_array(&mut f, "I_cot", &result.i_cot, true)?;
    writeln!(f, "  \"timestamp\": \"{}\"", timestamp)?;
    writeln!(f, "}}")?;

    Ok(())
}

pub fn write_results_csv(filepath: &str, result: &SimulationResult) -> std::io::Result<()> {
    let mut f = fs::File::create(filepath)?;
    writeln!(f, "Vsd_V,I_tol_A,I_seq_A,I_cot_A")?;
    for i in 0..result.vsd.len() {
        writeln!(
            f,
            "{:.6e},{:.6e},{:.6e},{:.6e}",
            result.vsd[i], result.i_tol[i], result.i_seq[i], result.i_cot[i]
        )?;
    }
    Ok(())
}

File: cpp/src/json_io.cpp
#include "json_io.hpp"

#include <cmath>
#include <cstdio>
#include <ctime>
#include <fstream>
#include <limits>

#include <nlohmann/json.hpp>

namespace fc {

SimParams parse_params_json(const std::string& filepath) {
    SimParams p;
    p.tau = std::numeric_limits<double>::infinity();

    std::ifstream f(filepath);
    if (!f.is_open()) {
        return p;
    }

    nlohmann::json root;
    try {
        root = nlohmann::json::parse(f);
    } catch (...) {
        return p;
    }

    auto& params = root["parameters"];
    auto& sweep = root["bias_sweep"];

    p.N = params.value("N", 0);
    p.vmode = params.value("vmode", 0.0);
    p.alphaL = params.value("alphaL", 0.0);
    p.alphaR = params.value("alphaR", 0.0);
    p.lambda = params.value("lambda", 0.0);
    p.T = params.value("T", 0.0);
    p.eta = params.value("eta", 0.0);
    p.Vg = params.value("Vg", 0.0);

    auto tau_val = params["tau"];
    if (tau_val.is_string() && tau_val.get<std::string>() == "Inf") {
        p.tau = std::numeric_limits<double>::infinity();
    } else if (tau_val.is_number()) {
        p.tau = tau_val.get<double>();
    }

    p.Vsd_start = sweep.value("Vsd_start", 0.0);
    p.Vsd_end = sweep.value("Vsd_end", 0.0);
    p.Vsd_step = sweep.value("Vsd_step", 0.0);

    return p;
}

std::vector<double> load_matlab_vsd(const std::string& filepath) {
    std::ifstream f(filepath);
    if (!f.is_open()) {
        return {};
    }

    nlohmann::json root;
    try {
        root = nlohmann::json::parse(f);
    } catch (...) {
        return {};
    }

    if (!root.contains("Vsd") || !root["Vsd"].is_array()) {
        return {};
    }

    return root["Vsd"].get<std::vector<double>>();
}

MatlabReference load_matlab_reference(const std::string& filepath) {
    MatlabReference ref;

    std::ifstream f(filepath);
    if (!f.is_open()) {
        return ref;
    }

    nlohmann::json root;
    try {
        root = nlohmann::json::parse(f);
    } catch (...) {
        return ref;
    }

    if (root.contains("I_tol") && root["I_tol"].is_array()) {
        ref.I_tol = root["I_tol"].get<std::vector<double>>();
    }
    if (root.contains("I_seq") && root["I_seq"].is_array()) {
        ref.I_seq = root["I_seq"].get<std::vector<double>>();
    }
    if (root.contains("I_cot") && root["I_cot"].is_array()) {
        ref.I_cot = root["I_cot"].get<std::vector<double>>();
    }

    return ref;
}

void write_results_json(const std::string& filepath, const std::string& spec,
                        const SimParams& params, const SimulationResult& res,
                        double wall_time) {
    FILE* fp = std::fopen(filepath.c_str(), "w");
    if (fp == nullptr) {
        return;
    }

    std::time_t now = std::time(nullptr);
    std::tm* lt = std::localtime(&now);
    char tbuf[64] = {0};
    if (lt != nullptr) {
        std::strftime(tbuf, sizeof(tbuf), "%Y-%m-%d %H:%M:%S", lt);
    }

    int n = static_cast<int>(res.Vsd.size());

    std::fprintf(fp, "{\n");
    std::fprintf(fp, "  \"spec\": \"%s\",\n", spec.c_str());
    std::fprintf(fp, "  \"language\": \"C++\",\n");
#if defined(__clang__)
    std::fprintf(fp, "  \"version\": \"clang %s\",\n", __clang_version__);
#elif defined(__GNUC__)
    std::fprintf(fp, "  \"version\": \"gcc %s\",\n", __VERSION__);
#else
    std::fprintf(fp, "  \"version\": \"unknown\",\n");
#endif
    std::fprintf(fp, "  \"wall_time_seconds\": %.17g,\n", wall_time);
    std::fprintf(fp, "  \"parameters\": {\n");
    std::fprintf(fp, "    \"N\": %d,\n", params.N);
    std::fprintf(fp, "    \"vmode\": %.17g,\n", params.vmode);
    std::fprintf(fp, "    \"alphaL\": %.17g,\n", params.alphaL);
    std::fprintf(fp, "    \"alphaR\": %.17g,\n", params.alphaR);
    std::fprintf(fp, "    \"lambda\": %.17g,\n", params.lambda);
    std::fprintf(fp, "    \"T\": %.17g,\n", params.T);
    std::fprintf(fp, "    \"eta\": %.17g,\n", params.eta);
    std::fprintf(fp, "    \"Vg\": %.17g,\n", params.Vg);
    if (std::isinf(params.tau)) {
        std::fprintf(fp, "    \"tau\": \"Inf\"\n");
    } else {
        std::fprintf(fp, "    \"tau\": %.17g\n", params.tau);
    }
    std::fprintf(fp, "  },\n");
    std::fprintf(fp, "  \"bias_sweep\": {\n");
    std::fprintf(fp, "    \"Vsd_start\": %.17g,\n", params.Vsd_start);
    std::fprintf(fp, "    \"Vsd_end\": %.17g,\n", params.Vsd_end);
    std::fprintf(fp, "    \"Vsd_step\": %.17g\n", params.Vsd_step);
    std::fprintf(fp, "  },\n");

    auto write_array = [&](const char* name, const std::vector<double>& arr, bool comma) {
        std::fprintf(fp, "  \"%s\": [\n", name);
        for (int i = 0; i < n; ++i) {
            std::fprintf(fp, "    %.17g%s\n", arr[i], (i + 1 < n) ? "," : "");
        }
        std::fprintf(fp, "  ]%s\n", comma ? "," : "");
    };

    write_array("Vsd", res.Vsd, true);
    write_array("I_tol", res.I_tol, true);
    write_array("I_seq", res.I_seq, true);
    write_array("I_cot", res.I_cot, true);
    std::fprintf(fp, "  \"timestamp\": \"%s\"\n", tbuf);
    std::fprintf(fp, "}\n");

    std::fclose(fp);
}

void write_results_csv(const std::string& filepath, const SimulationResult& res) {
    FILE* fp = std::fopen(filepath.c_str(), "w");
    if (fp == nullptr) {
        return;
    }

    int n = static_cast<int>(res.Vsd.size());
    std::fprintf(fp, "Vsd_V,I_tol_A,I_seq_A,I_cot_A\n");
    for (int i = 0; i < n; ++i) {
        std::fprintf(fp, "%.6e,%.6e,%.6e,%.6e\n",
                     res.Vsd[i], res.I_tol[i], res.I_seq[i], res.I_cot[i]);
    }

    std::fclose(fp);
}

}
